cmd/cloud-controller-manager/app/core.go

```diff
@@ -0,0 +1,111 @@
+		ctx.ClientBuilder.ClientOrDie("cloud-node-controller"),
+		cloud,
+		ctx.ComponentConfig.KubeCloudShared.NodeMonitorPeriod.Duration,
+		ctx.ComponentConfig.NodeStatusUpdateFrequency.Duration)
+
+	go nodeController.Run(stop)
+
+	return nil, true, nil
+}
+
+func startPersistentVolumeLabelController(ctx *cloudcontrollerconfig.CompletedConfig, cloud cloudprovider.Interface, stop <-chan struct{}) (http.Handler, bool, error) {
+	// Start the PersistentVolumeLabelController
+	pvlController := cloudcontrollers.NewPersistentVolumeLabelController(
+		ctx.ClientBuilder.ClientOrDie("pvl-controller"),
+		cloud,
+	)
+	go pvlController.Run(5, stop)
+
+	return nil, true, nil
+}
+
+func startServiceController(ctx *cloudcontrollerconfig.CompletedConfig, cloud cloudprovider.Interface, stop <-chan struct{}) (http.Handler, bool, error) {
+	// Start the service controller
+	serviceController, err := servicecontroller.New(
+		cloud,
+		ctx.ClientBuilder.ClientOrDie("service-controller"),
+		ctx.SharedInformers.Core().V1().Services(),
+		ctx.SharedInformers.Core().V1().Nodes(),
+		ctx.ComponentConfig.KubeCloudShared.ClusterName,
+	)
+	if err != nil {
+		// This error shouldn't fail. It lives like this as a legacy.
+		klog.Errorf("Failed to start service controller: %v", err)
+		return nil, false, nil
+	}
+
+	go serviceController.Run(stop, int(ctx.ComponentConfig.ServiceController.ConcurrentServiceSyncs))
+
+	return nil, true, nil
+}
+
+func startRouteController(ctx *cloudcontrollerconfig.CompletedConfig, cloud cloudprovider.Interface, stop <-chan struct{}) (http.Handler, bool, error) {
+
+	// If CIDRs should be allocated for pods and set on the CloudProvider, then start the route controller
+	if ctx.ComponentConfig.KubeCloudShared.AllocateNodeCIDRs && ctx.ComponentConfig.KubeCloudShared.ConfigureCloudRoutes {
+		if routes, ok := cloud.Routes(); !ok {
+			klog.Warning("configure-cloud-routes is set, but cloud provider does not support routes. Will not configure cloud provider routes.")
+		} else {
+			var clusterCIDR *net.IPNet
+			var err error
+			if len(strings.TrimSpace(ctx.ComponentConfig.KubeCloudShared.ClusterCIDR)) != 0 {
+				_, clusterCIDR, err = net.ParseCIDR(ctx.ComponentConfig.KubeCloudShared.ClusterCIDR)
+				if err != nil {
+					klog.Warningf("Unsuccessful parsing of cluster CIDR %v: %v", ctx.ComponentConfig.KubeCloudShared.ClusterCIDR, err)
+				}
+			}
+
+			routeController := routecontroller.New(
+				routes,
+				ctx.ClientBuilder.ClientOrDie("route-controller"),
+				ctx.SharedInformers.Core().V1().Nodes(),
+				ctx.ComponentConfig.KubeCloudShared.ClusterName,
+				clusterCIDR,
+			)
+			go routeController.Run(stop, ctx.ComponentConfig.KubeCloudShared.RouteReconciliationPeriod.Duration)
+		}
+	} else {
+		klog.Infof("Will not configure cloud provider routes for allocate-node-cidrs: %v, configure-cloud-routes: %v.", ctx.ComponentConfig.KubeCloudShared.AllocateNodeCIDRs, ctx.ComponentConfig.KubeCloudShared.ConfigureCloudRoutes)
```

Comment:
Should return `nil, false, nil` for this case here?

Reply:
IMO, no need. In this case, we just hint user in the logm but will still start the `RouteController`.

Reply:
I don't follow and maybe I missed something. Here we never actually run the route controller go routine but we will log "Started: routecontroller", that does not seem like the right behavior to me 🤔

Reply:
you are right. fixed now :+1: